### PR TITLE
Redesign user views

### LINKED ISSUE #40: Design User views
Show page should list user's appliances. Show page should work as an edit page as well (to set preferred floor like on the tanulo.sch.bme.hu Profile page).

### PR DESCRIPTION
**Pull request by triszt4n**
Closes #40 

### What's new:
* User view designs. :art: 
* Some fixes, see comments below. :bug:

### CODE REVIEW

**Comment by triszt4n** [comment]
Issue #60 could be solved with a new link on the user show page. (under the email address maybe)
But that issue needs advanced pék knowledge, I think.

**Comment by triszt4n** [comment]
This PR might even solve issue #55 but not on appliance show page (that wouldn't be smart), but on the kitchen show page AND on the user show page.

File: app/controllers/appliance_comments_controller.rb [hunk 1]
```diff
@@ -5,7 +5,7 @@ class ApplianceCommentsController < ApplicationController
   # GET /appliance_comments.json
   def index
     @appliance_comments = Appliance.includes(:appliance_comments).all
-                                   .map{ |appliance| appliance.comments.last }
+                                   .filter_map{ |appliance| appliance.comments.last if !appliance.comments.last.nil? }
                                    .sort_by(&:updated_at).reverse
   end
```
Comment (triszt4n) [inline app/controllers/appliance_comments_controller.rb:10]: It seems, if the appliance didn't have any comments, the last method would give back `nil`, with which `updated_at` in `sort_by` would always break the page from loading.

Comment (triszt4n) [reply app/controllers/appliance_comments_controller.rb:10]: The filter_map is a clever solution for getting only the existing comments.

File: app/controllers/appliances_controller.rb [hunk 2]
```diff
@@ -10,7 +10,7 @@ def show
 
   # GET /appliances/new
   def new
-    @appliance = Appliance.new
+    @appliance = Appliance.new(params.permit(:kitchen_id))
```
Comment (triszt4n) [inline app/controllers/appliances_controller.rb:13]: Making it possible to auto-fill in advance the kitchen field when creating new appliance.

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -9,6 +9,11 @@ def logged_in?
   end
   helper_method :logged_in?
 
+  def am_i_admin?
+    current_user.present? and current_user.admin
+  end
+  helper_method :am_i_admin?
+
```
Comment (triszt4n) [inline app/controllers/application_controller.rb]: Might be useful for later in view logics. (I've used it) If there's cleverer solution, let me know! :pray:

File: app/views/appliances/show.html.haml [hunk 4]
```diff
@@ -53,7 +53,8 @@
         .col-span-2.md:grid.md:grid-cols-2.py-1
           .text-sm.text-right.md:text-center
             = l comment.created_at, format: :short
-          .text-sm.font-medium.text-blue-600.text-righ= link_to comment.owner.name, user_path(comment.owner), class: "hover:underline"
+          .text-sm.font-medium.text-blue-600.text-right
+            = link_to comment.owner.name, user_path(comment.owner), class: "hover:underline"
```
Comment (triszt4n) [inline app/views/appliances/show.html.haml:57]: Yeah, this was a little mistyping accident that I corrected.

File: app/controllers/appliance_comments_controller.rb (hunk 1 above)
Comment (SepsiLaszlo) [reply app/controllers/appliance_comments_controller.rb:10]: Great solution! 
In this case I think the `filter_map` is the clearest implementation.
There is another way to remove `nil ` values from an array which could be useful in other cases. We could use the  [compact](https://ruby-doc.com/core/Array.html#method-i-compact) method to remove all `nil` values.

File: app/controllers/application_controller.rb (hunk 3 above)
Comment (SepsiLaszlo) [reply app/controllers/application_controller.rb]: This is an interesting problem.
We could check permissions in views based on roles with helper methods like you did which is very clear to read.
Or we use the polices for the corresponding action. This is not elegant, but it decouples the view from the authorization. (If some view's permission is changed we only have to change it in the policy not in all views.)
I will make a suggestion in the appropriate view.

File: app/views/layouts/_nav.html.haml [hunk 5]
```diff
@@ -12,6 +12,9 @@
         = link_to "Konyhák", kitchens_path, class: navLinkClass
       %li.py-1
         = link_to "Friss jelentések", appliance_comments_path, class: navLinkClass
+      - if am_i_admin?
```
Comment (SepsiLaszlo) [inline app/views/layouts/_nav.html.haml]: We could check if the action of the link is permitted for the `current_user` and render the button based on that.
Currently we could something like this:

But if we create a constructor for the `UserPolicy` with one argument we could use something like this:
```suggestion
      - if policy(current_user).index?
```
The constructor must accept both one and two arguments. One way to achive this is with a default argument.
The mentioned constructor in the `UserPolicy` for this to work:
```ruby
  def initialize(user, record = nil)
    super
  end
```

File: app/views/users/index.html.haml [hunk 6]
```diff
@@ -1,24 +1,21 @@
...
+  .px-6.py-3.bg-blue-500.text-left.text-sm.font-medium.text-white.uppercase{class: "w-1/2"}
+    Felhasználó
+  .px-6.py-3.bg-blue-500.text-right.text-sm.font-medium.text-white.uppercase{class: "w-1/2"}
+    Szint
+  - @users.each do |user|
+    = link_to user, class: "py-4 px-6 hover:bg-blue-200 transition-all duration-300 flex flex-wrap bg-white divide-gray-200 w-full" do
+      .text-sm{class: "w-3/4"}
+        .font-medium.text-gray-900
+          = user.name
+          - if user.admin
```
Comment (SepsiLaszlo) [inline app/views/users/index.html.haml:14]: I think this is fine, we can keep it that way. No need to use policy here.

File: app/views/users/show.html.haml [hunk 7]
```diff
@@ -1,18 +1,53 @@
-%p#notice= notice
+%h2.text-3xl.md:text-4xl.mb-4
+  %strong= @user.name
+  - if @user.admin
```
Comment (SepsiLaszlo) [inline app/views/users/show.html.haml:3]: We can keep that too.

File: app/views/users/show.html.haml [hunk 8]
```diff
@@ -1,18 +1,53 @@
...
-%p
-  %b Email:
-  = @user.email
-%p
-  %b Name:
-  = @user.name
-%p
-  %b Floor:
-  = @user.floor
+- if current_user == @user or am_i_admin?
```
Comment (SepsiLaszlo) [inline app/views/users/show.html.haml]: We could use a policy here too.
```suggestion
- if policy(current_user).update?
```

**Review by SepsiLaszlo — CHANGES_REQUESTED** [review]
Excellent work!
The views are very good to look at and they are perfectly ergonomic!

I left some suggestions on how we can better determine if some UI elements needs to be rendered or not.

**Comment by SepsiLaszlo** [comment]
There is a more compact way for using policies in views. I updated the suggestions.

**Review by SepsiLaszlo — APPROVED** [review]

**Pull request merged**